src/crypto/dtls.rs

```diff
@@ -0,0 +1,176 @@
+    Connected,
+
+    /// Keying material for SRTP encryption master key and the selected SRTP profile.
+    SrtpKeyingMaterial(KeyingMaterial, SrtpProfile),
+
+    /// The fingerprint of the remote peer.
+    ///
+    /// This should be checked against the fingerprint communicated in the SDP.
+    RemoteFingerprint(Fingerprint),
+
+    /// Decrypted data from incoming DTLS traffic.
+    Data(Vec<u8>),
+}
+
+/// Certificate used for DTLS.
+#[derive(Clone)]
+pub struct DtlsCert(DtlsCertInner);
+
+#[derive(Debug, Clone)]
+pub enum DtlsCertInner {
+    #[cfg(feature = "openssl")]
+    OpenSsl(super::ossl::OsslDtlsCert),
+}
+
+impl DtlsCert {
+    /// Create a new OpenSSL variant of the certificate.
+    #[cfg(feature = "openssl")]
+    pub fn new_openssl() -> Self {
+        let cert = super::ossl::OsslDtlsCert::new();
+        DtlsCert(DtlsCertInner::OpenSsl(cert))
+    }
+
+    /// Creates a fingerprint for this certificate.
+    ///
+    /// Fingerprints are used to verify a remote peer's certificate.
+    pub fn fingerprint(&self) -> Fingerprint {
+        match &self.0 {
+            #[cfg(feature = "openssl")]
+            DtlsCertInner::OpenSsl(v) => v.fingerprint(),
+            _ => unreachable!(),
+        }
+    }
+
+    pub(crate) fn create_dtls_impl(&self) -> Result<DtlsImpl, CryptoError> {
+        match &self.0 {
+            #[cfg(feature = "openssl")]
+            DtlsCertInner::OpenSsl(c) => Ok(DtlsImpl::OpenSsl(super::ossl::OsslDtlsImpl::new(
+                c.clone(),
+            )?)),
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl fmt::Debug for DtlsCert {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.0 {
+            #[cfg(feature = "openssl")]
+            DtlsCertInner::OpenSsl(c) => c.fmt(f),
+            _ => unreachable!(),
+        }
+    }
+}
+
+pub trait DtlsInner: Sized {
+    /// Set whether this instance is active or passive.
+    ///
+    /// i.e. initiating the client hello or not. This must be called
+    /// exactly once before starting to handshake (I/O).
+    fn set_active(&mut self, active: bool);
+
+    /// Handle the handshake. Once this succeeds, it becomes a no-op.
+    fn handle_handshake(&mut self, o: &mut VecDeque<DtlsEvent>) -> Result<bool, CryptoError>;
+
+    /// If set_active, returns what was set.
+    fn is_active(&self) -> Option<bool>;
+
+    /// Handles an incoming DTLS datagrams.
+    fn handle_receive(&mut self, m: &[u8], o: &mut VecDeque<DtlsEvent>) -> Result<(), CryptoError>;
+
+    /// Poll for the next datagram to send.
+    fn poll_datagram(&mut self) -> Option<DatagramSend>;
+
+    /// Handling incoming data to be sent as DTLS datagrams.
+    fn handle_input(&mut self, data: &[u8]) -> Result<(), CryptoError>;
+
+    /// Whether the DTLS connection is established.
+    fn is_connected(&self) -> bool;
+}
+
+pub enum DtlsImpl {
+    #[cfg(feature = "openssl")]
+    OpenSsl(super::ossl::OsslDtlsImpl),
+}
```

Comment:
What is the point of the `DtlsInner` trait if you are doing enum-based dispatch here? Isn't this even more LoC compared to having:

```rust
pub struct DtlsImpl(Box<dyn DtlsInner>);
```

YMMV but unless you have a `Box<dyn MyTrait>` or `T: MyTrait` somewhere, you don't actually need a trait so I'd argue you shouldn't have it.

Reply:
It's about establishing the contract. I find it easier to reason about the internal API when I see it like that. But it's not strictly necessary. It's following the pattern we have for Packetizer/Depacketizer.

Reply:
Dynamic dispatch is probably less performant than static too. We don't need the flexibility of dynamic dispatch here

Reply:
> Dynamic dispatch is probably less performant than static too. We don't need the flexibility of dynamic dispatch here

Yeah absolutely. I am not arguing for dynamic dispatch, I am arguing against using language features (traits) when they are not needed. Same reason as preferring regular functions over macros.

> It's about establishing the contract.

Aren't the public functions of a type / module a good enough contract? Other modules don't see the internal APIs anyway so I am not sure what exactly is achieved with this rigid enforcement other than adding layers of indirection when navigating the code.